### ISSUE TITLE
Proposal: Deprecate exports (except @JSExportTopLevel and @JSExportStatic)

### ISSUE THREAD

**Issue by sjrd** [scala-js/scala-js#2720]
**Edit:** The original proposal was to only deprecate `@JSExport` on instance members on Scala classes. The discussion showed that, once those are deprecated, there are little uses left for `@JSExport` on top-level classes and objects either. The use cases for those are real, but can be replaced with better alternatives, which were spun off as Proposals #2733 and #2734. The current text of the proposal includes those changes, and deprecates all uses of `@JSExport`, `@JSExportDescendantClasses`/`Objects`, and `@JSExportAll`.

Here is a potentially crazy idea: what if we deprecated (and removed) all exports except `@JSExportTopLevel` and `@JSExportStatic`?

Let us immediately see what exactly falls in that category:

## What does this proposal concern?

```scala
// Export members of Scala classes
object Foo {
  @JSExport def bar() = ???
  @JSExport val foobar = ???
  @JSExport var baz = ???
}

class Babar(@JSExport val x: Int)

// Export all members of a Scala class with @JSExportAll
@JSExportAll
case class CaseClass(x: Int, y: Int)

// Export a Scala class or object
@JSExport
class Foo
@JSExport
object Bar

// Non-top-level export of a JS class or object
@JSExport
@ScalaJSDefined
class Foo extends js.Object
@JSExport
@ScalaJSDefined
object Bar extends js.Object

// Export all descendants of a type (Scala or JS)
@JSExportDescendantClasses
trait SomeScalaTrait1
@JSExportDescendantObjects
trait SomeScalaTrait2

@JSExportDescendantClasses
@ScalaJSDefined
trait SomeJSTrait1 extends js.Object
@JSExportDescendantObjects
@ScalaJSDefined
trait SomeJSTrait2 extends js.Object
```

## What does this proposal *not* concern

```scala
// Export a JS class or object at the top-level
@JSExportTopLevel("Foo")
@ScalaJSDefined
class Foo extends js.Object

@JSExportTopLevel("Bar")
@ScalaJSDefined
object Bar extends js.Object

// Export object members at the top-level
object Container {
  @JSExportTopLevel("foo")
  def foo(): Int = 42
}

// Export object members as static in the companion class
@ScalaJSDefined
class Foo extends js.Object

object Foo {
  @JSExportStatic
  def foo(): Int = 42
}
```
In other words, all uses of `@JSExport`, `@JSExportDescendantClasses`, `@JSExportDescendantObjects` and `@JSExportAll` are deprecated. `@JSExportTopLevel` and `@JSExportStatic` are not deprecated.

The de facto "export" of all members of `@ScalaJSDefined` JS classes are not deprecated, obviously.

## Rationale

Let us start by rationalising why deprecating *instance* members of *Scala* classes, i.e., `@JSExport` on members of a Scala class/object, as well as `@JSExportAll`.

Once upon a time, when we were developing Scala.js 0.4.0, we had to come up with a way for JavaScript code to reliably call Scala.js code. That way was exports. The typical use case would look like
```scala
@JSExport("Main")
object Main {
  @JSExport
  def main(): Unit = { ... }
}
```
and that use case was even eventually codified into the trait `js.JSApp`, so people could simply do:
```scala
object Main extends js.JSApp {
  def main(): Unit = { ... }
}
```
without even having to specify any explicit `@JSExport`.

At the time, *Scala.js-defined JS classes did not exist*, nor did top-level exports (`@JSExportTopLevel`). Which means exporting one method somewhere in the codebase was the only way you could get Scala.js to run anything.

Now, we have Scala.js-defined JS classes, which are the most appropriate way to declare an API callable from JavaScript. The above use-case would be written as:
```scala
@JSExport("Main")
@ScalaJSDefined // <-- that won't be needed in 1.0.0
object Main extends js.Object {
  // visible from JS because Main is a JS type
  def main(): Unit = { ... }
}
```
And call it as `Main().main();` from JS, as before.

Even better, since 0.6.14, we can export functions at the top-level, which means we could simply do the following instead:
```scala
object Main {
  @JSExportTopLevel("main")
  def main(): Unit = { ... }
}
```
which we can call as just `main();` from JS.

If we had had Scala.js-defined JS classes from the beginning, it is not clear we would ever have allowed to export instance members of Scala classes. Even more so with `@JSExportTopLevel` in addition.

## Use cases and replacements

Let's examine the use cases for `@JSExport`ed members, and what their (arguably better) replacements are in this day and age:

### Entry point of the application

As mentioned above, the Number 1 use case for exports is the main entry-point of an application. We can use an `@JSTopLevelExport` function with a better JS API, now.

For source compatibility (and to keep sbt "launcher" auto-detection), we can still define `js.JSApp` as follows without `@JSExport`ed member:
```scala
@ScalaJSDefined
trait JSApp extends js.Object {
  def main(): Unit
}
```

### Write a JavaScript library in Scala.js

A more involved use case of `@JSExport` is to create, in Scala.js, a library intended to be used from a JavaScript codebase. See for example [How to make an idiomatic JavaScript library with Scala.js](http://pavkin.ru/how-to-make-an-idiomatic-javascript-library-with-scala-js/).

In this use case, JavaScript code is definitely intended to manipulate objects defined in Scala.js on a regular basis, and to have access to their entire API. I believe that this is perfectly addressed by Scala.js-defined JS classes. For example, instead of:
```scala
@JSExport
@JSExportAll
class MyLibrary {
  def inc(x: Int): Int = x + 1
  def dec(x: Int): Int = x - 1
}
```
we would offer:
```scala
@JSExport
@ScalaJSDefined
class MyLibrary extends js.Object {
  def inc(x: Int): Int = x + 1
  def dec(x: Int): Int = x - 1
}
```

### case classes whose fields can be manipulated from JS

A quite common use case of exported members is to export the fields of a `case class`. It is often combined with exporting (the constructor of) the case class, for example:
```scala
@JSExport("Point")
case class Point(@JSExport x: Int, @JSExport y: Int)
```
or, equivalently,
```scala
@JSExport("Point")
@JSExportAll
case class Point(x: Int, y: Int)
```
This is a quite convenient construct. It allows both to create and read `Point`s from JS:
```js
var p = new Point(5, 6);
console.log(p.x);
console.log(p.y);
```
as well as idiomatically manipulate `Point`s from Scala, all with pattern matching, structural equality, and things like that.

To be completely honest, there is no direct replacement for this use case. The closest one can get is a Scala.js-defined JS class, with hand-written `apply()` and `unapply()` if necessary, like this:
```scala
@JSExport("Point")
@ScalaJSDefined
class Point(val x: Int, val y: Int) extends js.Object {
  def toString(): String = s"Point($x,$y)"
}

object Point {
  def apply(x: Int, y: Int): Point = new Point(x, y)
  def unapply(p: Point): Some[(Int, Int)] = Some((p.x, p.y))
}
```
*However*, such a `Point` does *not* have structural equality. As any other JS type, `==` tests for reference equality (`eq`) and is not overridable. In other words, `Point(1, 2) != Point(1, 2)` because they are not `eq`.

This can be annoying to work around. The most sensible thing to do would be to have a `case class Point` and a `class JSPoint extends js.Object`, with implicit conversions between them:
```scala
implicit def point2pointJS(p: Point): PointJS = PointJS(p.x, p.y)
implicit def pointJS2point(p: PointJS): Point = Point(p.x, p.y)
```
But that does not scale to collections of such points, or things like that.

If retaining mutability is desired, we can instead use a JS "view" object like the following. This can also be used in the immutable case, but is less critical then:
```scala
class Point(var x: Int, var y: Int) {
  val jsView: js.Object = new js.Object {
    def x: Int = Point.this.x
    def x_=(v: Int): Unit = Point.this.x
    def y: Int = Point.this.y
    def y_=(v: Int): Unit = Point.this.y
  }
}
```
If this pattern turns out to be common, it can be turned into a blackbox def macro in a library.

## Why? aka Problems with `@JSExport` on instance members

OK but ... why do you want to get rid of instance member exports!? They're fine, and we do use them. Don't just take away from us for the only reason that we (sort of) have alternatives!

Well, there are a few problems with `@JSExport` for member values in Scala classes.

### Arbitrary restrictions

`@JSExport`ed members suffer from arbitrary restrictions. The most glaring being that their export name cannot contain `__` (double-underscore). Why? Because the mangling of non-exported methods contains `__` (e.g., `def foo(x: Int): String` becomes `foo__I__T` in JS), and this is how the implementation makes sure that exported methods do not clash with non-exported in JavaScript.

That's stupid. A language feature should not be constrained in arbitrary ways justified by the underlying implementation.

### Implementation hazard

Exported instance members are actually quite hard to implement. Yes, we've done the work, so we could keep it like that, but it leaves traces. For example, to implement them, we have to generate synthetic (Scala) methods with mangled names in the classes. They can cause pretty cryptic error messages like the following:
```
[error] HelloWorld.scala:15: double definition:
[error] def $js$exported$meth$bar(): Any at line 12 and
[error] def $js$exported$meth$bar(): Any at line 15
[error] have same type
[error]   @JSExport("bar")
[error]    ^
[error] one error found
```
caused by this code:
```scala
class Foo {
  @JSExport
  def bar(): Int = 42

  @JSExport("bar")
  def foo(): Int = 5
}
```

### Potentially incompatible with a compilation to WebAssembly

Last but not least, the existence of instance member exports is potentially putting in jeopardy the future ability to compile Scala.js to WebAssembly. At the moment, wasm lacks enough features to compile any significant subset of Scala.js to it. But in the future, it is likely that it will eventually receive (typed) heap objects, and the ability to interoperate with JavaScript libraries, as Scala.js can. However, I expect it will not allow wasm typed objects to define methods that can be called from JavaScript. If that is the case, the existence of exported members in Scala classes would prevent us from compiling Scala classes to wasm "classes" (whatever that means). Instead, we would have to only emit "JavaScript" classes, and only the content of methods could be implemented in wasm (but not the memory layout, virtual dispatch, etc.).

Note that is speculation. This is my educated guess, based on discussions happening on https://github.com/WebAssembly/design and on informal discussions I had with some people deeply involved in wasm.

## Other deprecations

So now that we have made a case to deprecate exports of instance members of Scala classes, let's see the other uses of exports.

### Exporting Scala classes and objects

First, exporting a Scala class or object becomes kind of moot. The object/class itself would be exported, but once you get it in JS, there's nothing you can actually *do* with it, other than pass it around and back to Scala.js code. Therefore, it seems like it makes sense to deprecate `@JSExport` on Scala classes and objects (but not on JS classes and objects). This also includes `@JSExportDescendantClasses` and `@JSDescendantObjects`.

However, there is still one use case for the latter two: they are used by testing frameworks to "reflectively" instantiation test classes. We cannot wave that use case away, or we wouldn't have testing frameworks any more!

That said, using exports for this is basically an abuse. We use that because there is no other way to instantiate a class by its name. We can provide a better, disciplined API for this use case, which is Proposal #2733. With that API in place, testing frameworks (and Akka.js) do not need `@JSExportDescendantClasses` nor `@JSDescendantObjects` any more.

This allows to safely deprecate exports of Scala classes and objects.

### Exporting JS classes and objects

Exporting JS classes and objects is important. We won't remove that. However, the way `@JSExport` works is not very nice for objects, because they are exported as functions returning the object, rather than the object itself. Proposal #2743 introduces new uses of `@JSExportTopLevel` to replace `@JSExport` on JS classes and objects, so that objects are directly exported.

This rules out the last remaining use of `@JSExport`.

## Thoughts, objections?

Let us know in the comments below! (as some Youtubers like to say :-p)

**Comment by jducoeur** [comment]
Taking a quick look at my own usage, I think I'm comfortable with this, provided that there's a decent deprecation period.  It would affect more or less all of my `@JSExport`s, but I think they'd all be straightforward to fix as described above.  I don't think I'm using the problematic `case class` example, although that does sound like a non-trivial concern for some use cases...

**Comment by gzm0** [comment]
First of all, as someone only hacking on the compiler and not actually using it very often, I'm of course all in favor of that.

A couple of thoughts.

# Time Line with respect to 1.0.0
Saying exports are gone in 1.0.0 seems pretty harsh. However, if we don't remove them in 1.0.0, we will be stuck with them for a while (if we want to guarantee backwards compatibility). So deprecating them is nothing more than a strong recommendation.

We (i.e. the ones hacking the compiler) would of course want this rolled out ASAP, so we don't need to support the whole thing. Maybe I have to ask back @jducoeur what order of magnitude (in days :P) is a decent deprecation period?

# Export typechecker (aka `js.use(x).as[T]`)
We could (and should) get rid of this thing as well, since we'd bring the typesystem back into a state where nominal typing is enough. This is nice, since it is probably impossible to support symbol based fields with the export typechecker.

# Exporting plain Scala classes
Why would we allow exporting plain Scala classes (and objects)? It seems the only halfway valid use case would be to pass one into an exported method. Also would that not be a problem for wasm?

**Comment by gzm0** [comment]
Also: We could allow Scala.js defined case classes, couldn't we? Is there something fundamental preventing us from it?

**Comment by jducoeur** [comment]
<i>Maybe I have to ask back @jducoeur what order of magnitude (in days :P) is a decent deprecation period?</i>

It's a good question, and I'll admit that I expected this objection.  I'm honestly not sure.  Certainly at least a couple of weeks, so that people can try doing without and make sure there aren't any deathly-important use cases that aren't anticipated above -- this *is* a commonly-used and fairly important feature now, so it's important to make sure that there are no critical issues with this API change.

Beyond that, I dunno.  *I* could cope with a deprecation period of a couple of weeks, but of course I'm running a little one-man shop so I'm fairly agile.  I'd recommend poking folks on Gitter for other opinions, and see if anybody squacks.  (I'm afraid that, even with @sjrd pointing out this Issue, not too many appear to actually venture over here.  Changing that to "we *are* planning to make this change, so please try it out *right now*" might get more attention.)

As to `use().as()`, I haven't found a need for it myself, so no objection here.  You might see whether anybody objects...

**Comment by gzm0** [comment]
> least a couple of weeks

Ok. So that would definitely be doable. Just wanted to have a ballpark :)

**Comment by sjrd** [comment]
> Time Line with respect to 1.0.0

Agreed. If we're not going to remove it by 1.0.0, there's no point deprecating it. My suggestion would be deprecate it in 0.6.15 (assuming we get consensus about that before, say, the end of January), keep it available but deprecated in 1.0.0-M1 or M2 and remove it by M3 or at the latest by the first RC.

This would be something like a 2-3 months deprecation period, including at least one 0.6.x version and at least one milestone of 1.0.0.

> Export typechecker

Indeed. We can easily move it to an external library if we're not comfortable getting rid of it entirely.

> Exporting plain Scala classes

This crossed my mind. In theory there shouldn't be any real *export* use case for exporting Scala classes/objects anymore. However, in practice, this is *the* way testing frameworks get to instantiate test classes "reflectively". We *have* to find a good replacement for that use case.

Note that this "reflective instantiation" is the reason we need `environmentInfo.exportsNamespace`. And *that* thing is going to be a serious pain in the neck when we'll emit ES6 modules. ES6 modules do not have a notion of object with all exports within the same module. You can see it from *another* module A that imports B with a namespace import `import * from 'B.js'`. But A itself cannot see its own exports. That is another good reason to find a good replacement for how testing frameworks get to instantiate their test classes.

> We could allow Scala.js defined case classes, couldn't we? Is there something fundamental preventing us from it?

Fundamental reason: we cannot override `equals` and `hashCode` in Scala.js-defined JS classes. Hence we cannot implement structural equality for Scala.js-defined case classes.

**Comment by sjrd** [comment]
@jducoeur 

>  (I'm afraid that, even with @sjrd pointing out this Issue, not too many appear to actually venture over here. Changing that to "we are planning to make this change, so please try it out right now" might get more attention.)

I plan to post regular reminders with increasing levels of urgency about this issue.

**Comment by gzm0** [comment]
> testing frameworks get to instantiate test classes "reflectively"

Ah, ah. Good point. I'll think about that :)

**Comment by mkotsbak** [comment]
Well, case classes are very useful part of Scala, so they need to be supported somehow (still available as scala defined after 1.0?). I suspect in many use cases the equals and hashcode of the objects are only used in the scala part of the code (where they could be named something else, a new "magic" method name) and not relied on by js libs that gets the objects passed to them, and then probably should be possible to explicitly enabled by the user.

**Comment by sjrd** [comment]
> Well, case classes are very useful part of Scala, so they need to be supported somehow (still available as scala defined after 1.0?).

Scala.js-defined JS case classes have never been supported. Normal case classes (not extending `js.Any`) are of course not going anywhere.

> I suspect in many use cases the equals and hashcode of the objects are only used in the scala part of the code (where they could be named something else, a new "magic" method name) and not relied on by js libs that gets the objects passed to them, and then probably should be possible to explicitly enabled by the user.

That's an interesting take on the issue. I'll have to think about it.

**Comment by andreaTP** [comment]
my two cents, not sure if you will take this in consideration since I'm relying on hacks of course...
but the proposed replacement
```scala
@JSExport("Point")
@ScalaJSDefined
class Point(val x: Int, val y: Int) extends js.Object {
  def toString(): String = s"Point($x,$y)"
}
```
would break much of the effort in place for Akka.Js.
Two critical points are:
1- class signature change
2- explicit name in JSExport

right now I'm patching code at compile time from a configuration file (i.e. https://github.com/akka-js/akka.js/blob/master/config/annotation_adder.config#L2 ) but it looks to me that this change will require a really deep rework of this....

Hope I'm missing something that keep this easily affordable...

**Comment by sjrd** [comment]
@andreaTP I think what you're missing is that we're only discussing `@JSExport`ed *instance members*. AFAICT Akka.js only uses `@JSExport` annotations *on classes and objects*.

@gzm0 suggested also removing `@JSExport` on classes that do not extend `js.Any`, which *would* affect your use case. But I said that if we do that we need a replacement for testing frameworks to be able to "reflectively" instantiate classes. Akka.js would be able to use that very same replacement with exactly as much effort as now.

> explicit name in JSExport

It's only explicit in this example so that the JS code does not see package names. The variant with an implicit name is not going away.

**Comment by andreaTP** [comment]
@sjrd Akka.Js make use of ```@JSExport``` for instance members.
i.e. ```self``` and ```context```

**Comment by sjrd** [comment]
Hum ... Can you show me where you *use* the exported `self` and/or `context`?

**Comment by andreaTP** [comment]
sure:
https://github.com/akka-js/akka.js/blob/master/akka-js-actor/js/src/main/scala/akka/util/Reflect.scala#L36

**Comment by sjrd** [comment]
That's a so-called reflective call (call of a member of a structural type). It has nothing to do with exports. In other words, you do not need to export `self` for this to work.

**Comment by sjrd** [comment]
To address reflective instantiation in testing frameworks, I opened a separate proposal #2733.

**Comment by sjrd** [comment]
@mkotsbak @gzm0 In fact, there is another fundamental problem with case classes: they extend `Product` (therefore `Equals`) as well as `Serializable`. But those interfaces do not extend `js.Any`, and a JS class cannot extend non-JS traits like that. I don't think there is a solution to this problem under the constraint that the typechecker cannot be changed.

**Comment by sjrd** [comment]
I have updated the text of the proposal to reflect the introduction of #2733 and #2743, and hence the deprecation of *all* uses of `@JSExport`, `@JSExportDescendantClasses`, `@JSExportDescendantObjects` and `@JSExportAll`.

**Comment by jtcwang** [comment]
At my company we use Scala.js to share core business logic between the browser (consumed by JS) and the Scala backend. We have a lot of exported classes and we use `@JSExport` extensively to avoid needing to define JS wrapper classes around our Scala classes. For example:

```
class Service {

  def fetchMetadata: Future[Metadata] = {
    //   ...
  }

  /** Re-export fetchMetadata for JS compatibility. */
  @OnlyJS @JSExport("fetchMetadata")
  def js_fetchMetadata: Future[Metadata] = {
    toJS(fetchMetadata(schema)
  }

}
```
Similarly, the `Metadata` is a normal Scala class with some `@JSExport` sprinkled on js_* to access its fields. Will there be a replacement for this use case?

**Comment by sjrd** [comment]
@jtcwang I think for this use case you can use an approach similar to the one presented in the proposal text for mutable classes, i.e., using a proxy JS "view", member of the `Service` class (and of `Metadata`). For `Service` it would look like the following.
```scala
class Service {

  def fetchMetadata: Future[Metadata] = {
    //   ...
  }

  /** Re-export fetchMetadata for JS compatibility. */
  @OnlyJS
  val jsView = new js.Object {
    def fetchMetadata: Future[js.Object] =
      Service.this.fetchMetadata.map(_.jsView) // assuming Metadata also has its jsView
  }
}
```
It doesn't seem to be more boilerplate than your original code.

**Comment by jimmydivvy** [comment]
In my organization we've made pretty heavy use of JSExport on instance members in libraries we export to JS, and there's a few scenarios that I'm not sure will be easily covered by the proposed solution. 

To give a brief overview of our architecture - we write most of our core business logic in Scala as libraries, and then call this from both frontend UIs written in Javascript, as well as JVM codebases on the server. Having a fast and low-overhead FFI between ScalaJS and Javascript is one of the things we really like about ScalaJS, and it would be a shame to limit this.


### Added boilerplate

From the original post: 

> A more involved use case of @JSExport is to create, in Scala.js, a library intended to be used from a JavaScript codebase. See for example How to make an idiomatic JavaScript library with Scala.js.
> In this use case, JavaScript code is definitely intended to manipulate objects defined in Scala.js on a regular basis, and to have access to their entire API. *I believe that this is perfectly addressed by Scala.js-defined JS classes.*

This makes sense if you're writing a Javascript library in Scala, however it's not really as viable an alternative if you're writing a Scala library to be used across both JVM and JS codebases. For us, a major benefit of ScalaJS is to allow code sharing, and we have a fairly significant codebase that can be used across with JS and JVM environments with very little boilerplate.

At the moment you can write Scala code *once* and reuse across both architectures almost out of the box without needing to write a separate facade for Javascript. You write it once, and apart from a few edge cases it just works whether we're calling a global static method or an instance method of a model class four layers deep. (The Scala.js deserve some serious kudos for how well it works).

Writing facades/wrappers for each of these is hypothetically possible, but it adds a lot of developer overhead to maintain and likely increased JS bundle sizes. 

Even if we could autogenerate this using macros, we'd still have a second problem - performance.

### Performance overhead 

We pass state between library and application code quite frequently, and manipulate it on both sides. A model might get passed back and forwards many times between layers as part of a single transaction. At the moment, we can just pass a reference around with negligable performance overhead as it crosses the Scala and JS boundaries, since the same instance with the same methods can be called on both sides. No conversion or wrapping needed. 

Removing JSExport and creating facade wrappers means every time you pass something between layers would require boxing or unboxing at best, and recursive conversion at worst. As an extreme example, we currently have UI components that render large sets of data at 60fps with JS -> ScalaJS calls as part of the rendering logic. The extra boxing overhead would likely make this impossible.

We also make use of reference equality for performance - we use ReactJS and assume all models are immutable, therefore a diff to check if a component needs rerendering is as simple as checking reference equality of the root node. Reboxing after crossing the FFI boundary would break this.


tl;dr 
- Significantly more boilerplate
- Increased JS bundle size
- Additional performance overhead when calling Scala from JS 

### Alternatives

The points raised in the original post do make sense, and I'm certainly not against updating my code if this needs changing. However the proposed solution would, at least for our use-case, likely require a major refactoring of a fairly large codebase and I'd like to at least open the discussion to try and find alternate solutions that provide a similarly powerful FFI as is available today. 


I'd like to also mention the following blog post: https://www.scala-lang.org/news/2015/02/05/scala-js-no-longer-experimental.html

> The semantics of the language are settled. They may still evolve in a backward compatible way (defining previously undefined behaviors), but will otherwise not change anymore.

> The standard library of Scala.js (the scala.scalajs.js package) will remain backward source and binary compatible.

> The sbt builds will remain backward source compatible.

While I can understand that sometimes these kinds of breaking changes do need to be made, there was a somewhat explicit claim of Scala.js being production ready with an intend to maintain backwards compatibility. As an organization that has made a substantial investment in Scala.js based on this assertion, we would very much like to encourage the ScalaJS team to keep this discussion open for a little while longer to make sure any kinks can be ironed out before comitting.

**Comment by sjrd** [comment]
@jimmydivvy

Some meta-comments first:

> I'd like to at least open the discussion to try and find alternate solutions that provide a similarly powerful FFI as is available today

The discussion is definitely open. That's the point of writing a proposal like this, and getting it as much visibility as possible. If it turns out the community is against it, we'll back it off. It would not be the first time that the community prevents us from removing support for something: we could not drop support for JDK6 for that reason.

> I'd like to also mention the following blog post: https://www.scala-lang.org/news/2015/02/05/scala-js-no-longer-experimental.html

Fair enough. Although, to my/our defence, when we wrote this, we thought 1.0.0 was going to be released a few months afterwards. In the meantime, we have added to Scala.js 0.6.x, while maintaining not only source but also binary compatibility, features that, back then, I dreamed about for a Scala.js **2**.0.0: Scala.js-defined JS classes, `@JSImport`, and many other smaller things. In a sense, we have so far exceeded my expectations in bringing improvements while maintaining compatibility.

---

On to the technical discussion.

You are raising a seriously good point. If I try to summarize what's unique about your use case, and that I not foreseen, is the combination of:
* Cross-platform libraries, directly called both from Scala on the server and JS on the client.
* Hard requirements on the performance of JavaScript-to-Scala calls.

If it were only cross-platform, the `jsView` pattern would do. And if it were only the performance aspect, then a Scala.js-defined JS class API would fit the requirements. But it seems hard to address both requirements without `@JSExport` of instance members, indeed.

Right now, I think this use case kills the proposal. I will think more about this to see if I can find a solution that would address those requirements without being as internally fragile/complicated as exports of instance members.

**Comment by gzm0** [comment]
I'd like to chime in here as well and draw a parallel to Scala on the JVM. Scala on the JVM is nice to use, because it tries to keep the gap between Java, the main language on its execution platform, as small as possible.
This is a trend that continues on (e.g. `@static`) and as far as I can tell, something much appreciated as it without doubt minimizes friction between Scala and Java.

This proposal would make Scala.js move in the opposite direction. The equivalent on Scala JVM, introducing "Java-style classes", sounds absurd. Despite the interaction not being perfect (e.g. value classes, varargs, default parameters) it seems to be simply "good enough".

Maybe we were thinking too strongly "inside the compiler" this time?

This makes me think: Maybe we should try and see if we can push in the other direction and make **all** classes Scala.js defined JavaScript classes (with some opt-in for visibility from JS) and emit what we call a Scala class today only as an "optimization".

**Comment by sjrd** [comment]
> Maybe we should try and see if we can push in the other direction and make all classes Scala.js defined JavaScript classes (with some opt-in for visibility from JS) and emit what we call a Scala class today only as an "optimization".

No, the semantics of Scala.js-defined JS classes are such that they kill all attempts at optimizations. No inlining, no stack allocation, no nothing. Besides, the rules for overloading are incompatible, which would break valid Scala code.

**Comment by lihaoyi** [comment]
> No, the semantics of Scala.js-defined JS classes are such that they kill all attempts at optimizations. No inlining, no stack allocation, no nothing. Besides, the rules for overloading are incompatible, which would break valid Scala code.

Should this be written up somewhere in the proposal at the top of this issue?

You touch on some limitations later on but it would be nice to have them all listed out so we can compare the `@JSExport` limitations with the Scala.js-defined classes limitations. After all, the proposal is basically replacing a lot of `@JSExport`s out in the wild with Scala.js-defined classes right?

**Comment by sjrd** [comment]
Note that, when called from JS (which is the goal of exports), all the same semantics apply to `@JSExport`. But it's true that, when called from Scala, even an `@JSExport`ed method can be inlined, for example.

**Comment by jimmydivvy** [comment]
Thanks Sébastien. We're by no means attached to JSExport as it stands today and we're not against breaking changes, so don't feel like my prior comment is pressure to religiously maintain backwards compatibility. Some of our team have had some interesting ideas on tweaking the FFI in the past, so I might see if we can put some of their notes together to further the discussion and to see if we can help contribute in any way.

For now, I've played around with a few examples from our codebase and tried to boil them down to a simple "stress test". I think the below contrived example probably highlights some of potential gotchas that may need to be considered as part of an FFI changes. 
   
	trait CanAdd[T <: CanAdd[T]] {
	    def add( other:T ):T
	}

	trait HasX {
        def x:Int
	}

	trait HasY {
	    def y:Int
	}

	@JSExportAll
	case class Vector(x:Int, y:Int) extends CanAdd[Vector] with HasX with HasY {
		def add( other:Vector ):Vector = Vector( other.x + this.x, other.y + this.y )
	}


While in this example the 'add' function is trivial and could just be copy pasted, assume for sake of argument that's a large complex set of logic that we want to reuse across platforms.

For this example - without the ability to export individual types to both platforms I don't see any way to implement this as cross platform code without effectively maintaining two completely separate copies of most of this code. The only thing that could really be shared would be the business logic inside `add` itself, and even that would require some fiddly conversion code.

I'll have a chat with our team next week and see if anyone has any further ideas.

**Comment by sjrd** [comment]
Well ... I'm been thinking about this on and off for the past few days, and I don't think there is any way to salvage this proposal.

Let's close it as rejected.
